app/jobs/submissions_job.rb

```diff
@@ -5,6 +5,101 @@ def add_warning_messages(messages)
     Rails.logger.error msg
   end
 
+  def copy_old_grading_data(new_submission, grouping)
+    # at this point, it will only have one result
+    new_result = new_submission.current_result
+
+    # get the last submission that wasn't the one we just created
+    old_submission = grouping.submissions.where.not(id: new_submission.id)
+                             .order(created_at: :desc)
+                             .first
+
+    # no old submissions => this option does nothing
+    return if old_submission.nil?
+
+    old_result = old_submission.non_pr_results
+                               .where(remark_request_submitted_at: nil)
+                               .last
+
+    result_data = [
+      { old: old_result.annotations, new: new_result.annotations },
+      { old: old_result.marks, new: new_result.marks },
+      { old: old_result.extra_marks, new: new_result.extra_marks }
+    ]
+
+    # copy over data from old result to new result
+    result_data.each do |result_set|
```

Comment:
Same copy logic for annotations, marks, extra marks.

Reply:
Overall, there are enough differences in behaviour that I don't think this loop is worth it.

1. Only existing _marks_ need to be destroyed (this is the only model with some pre-existing objects)
2. Only _annotations_ need to have a separate check for an existing submission file (which you've correctly noted below may trigger a case where an annotation _isn't_ copied over).
3. For _extra marks_, you should copy over all point-based marks (these are set manually by the grader), but _ignore_ all percentage-based marks (these are set by the late penalty, which is something that the instructor decides to apply, or not, for the newly-collected submission). Please make sure to document the latter case, as it's a bit subtle.

Reply:
This has also been fixed, and I've added docs both within the code and on the Wiki!